File: quizzes/quizzesV2/quizzes10/03-problem.ts
export {};

// @ts-ignore
const n: number = Number(process.argv[3]) || 2;

type Product = {
    name: string;
    quantity_sold: number;
};

const products: Product[] = [
    { name: 'Apples', quantity_sold: 120 },
    { name: 'Oranges', quantity_sold: 75 },
    { name: 'Bananas', quantity_sold: 200 },
];

const findTopSoldProductsWithSort = (): Product[] => {
    return products
        .toSorted((a, b) => b.quantity_sold - a.quantity_sold)
        .slice(0, n);
};

const findTopSoldProductsWithLoop = (): Product[] => {
    const topSoldProducts: Product[] = [];

    for (const product of products) {
        if (topSoldProducts.length < n) {
            topSoldProducts.push(product);
            topSoldProducts.sort((a, b) => b.quantity_sold - a.quantity_sold);
        } else if (
            product.quantity_sold > topSoldProducts[n - 1].quantity_sold
        ) {
            topSoldProducts[n - 1] = product;
            topSoldProducts.sort((a, b) => b.quantity_sold - a.quantity_sold);
        }
    }

    return topSoldProducts;
};

console.log(JSON.stringify({ data: findTopSoldProductsWithSort() }, null, 2));
console.log(JSON.stringify({ data: findTopSoldProductsWithLoop() }, null, 2));

File: bins/binsV2/bin/bins2/03-problem.rb
# frozen_string_literal: true

require "json"

n = (ARGV[3] || 2).to_i

products = [
  { name: 'Apples',  quantity_sold: 120 },
  { name: 'Oranges', quantity_sold: 75  },
  { name: 'Bananas', quantity_sold: 200 }
]

def self.find_top_sold_with_sort(products, n)
  products.sort_by { |p| -p[:quantity_sold] }.first(n)
end

def self.find_top_sold_with_loop(products, n)
  top_sold = []

  products.each do |product|
    if top_sold.length < n
      top_sold << product
      top_sold.sort_by! { |p| -p[:quantity_sold] }
    elsif product[:quantity_sold] > top_sold.last[:quantity_sold]
      top_sold[-1] = product
      top_sold.sort_by! { |p| -p[:quantity_sold] }
    end
  end

  top_sold
end

puts JSON.pretty_generate({ data: find_top_sold_with_sort(products, n) })
puts JSON.pretty_generate({ data: find_top_sold_with_loop(products, n) })
